validator/client/attest_protect.go

```diff
@@ -106,31 +130,63 @@ func isNewAttSlashable(history *slashpb.AttestationHistory, sourceEpoch, targetE
 	if history == nil {
 		return nil
 	}
 	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
-
-	if targetEpoch > history.LatestEpochWritten {
+	lew, err := history.GetLatestEpochWritten(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get latest epoch written from encapsulated data")
+		return nil
+	}
+	if targetEpoch > lew {
 		// If the target epoch to mark is ahead of latest written epoch, override the old targets and mark the requested epoch.
 		// Limit the overwriting to one weak subjectivity period as further is not needed.
-		maxToWrite := history.LatestEpochWritten + wsPeriod
-		for i := history.LatestEpochWritten + 1; i < targetEpoch && i <= maxToWrite; i++ {
-			history.TargetToSource[i%wsPeriod] = params.BeaconConfig().FarFutureEpoch
+		maxToWrite := lew + wsPeriod
+		for i := lew + 1; i < targetEpoch && i <= maxToWrite; i++ {
+			history, err = history.SetTargetData(ctx, i%wsPeriod, &kv.HistoryData{Source: params.BeaconConfig().FarFutureEpoch})
+			if err != nil {
+				log.WithError(err).Error("Could not set target to the encapsulated data")
+				return nil
+			}
+		}
+		history, err = history.SetLatestEpochWritten(ctx, targetEpoch)
+		if err != nil {
+			log.WithError(err).Error("Could not set latest epoch written to the encapsulated data")
+			return nil
 		}
-		history.LatestEpochWritten = targetEpoch
 	}
-	history.TargetToSource[targetEpoch%wsPeriod] = sourceEpoch
+	history, err = history.SetTargetData(ctx, targetEpoch%wsPeriod, &kv.HistoryData{Source: sourceEpoch, SigningRoot: signingRoot[:]})
+	if err != nil {
+		log.WithError(err).Error("Could not set target to the encapsulated data")
+		return nil
+	}
 	return history
 }
 
 // safeTargetToSource makes sure the epoch accessed is within bounds, and if it's not it at
-// returns the "default" FAR_FUTURE_EPOCH value.
-func safeTargetToSource(history *slashpb.AttestationHistory, targetEpoch uint64) uint64 {
+// returns the "default" nil value.
+func safeTargetToSource(ctx context.Context, history *kv.EncHistoryData, targetEpoch uint64) *kv.HistoryData {
 	wsPeriod := params.BeaconConfig().WeakSubjectivityPeriod
-	if targetEpoch > history.LatestEpochWritten || int(targetEpoch) < int(history.LatestEpochWritten)-int(wsPeriod) {
-		return params.BeaconConfig().FarFutureEpoch
+	lew, err := history.GetLatestEpochWritten(ctx)
+	if err != nil {
+		log.WithError(err).Error("Could not get latest epoch written from encapsulated data")
+		return nil
+	}
+	if targetEpoch > lew {
+		return nil
+	}
+	if lew >= wsPeriod && targetEpoch < lew-wsPeriod { //Underflow protected older then weak subjectivity check.
+		return nil
+	}
+	hd, err := history.GetTargetData(ctx, targetEpoch%wsPeriod)
+	if err != nil {
+		log.WithError(err).Errorf("Could not get target data for target epoch: %d", targetEpoch)
+		return nil
+	}
+	if hd.IsEmpty() {
```

Comment:
this check should be performed by the caller of the function, return nil here makes no difference in the event `hd` is nil anyway.

Reply:
thanks